### ISSUE TITLE
Comparison to numbalsoda

### ISSUE THREAD

**Issue by jdeseze** [patrick-kidger/diffrax#577]
Hi,
I am solving complex biochemical equations with lots of parameters.  
I was using numbalsoda for fast solving of my equations.
I'd love to switch to diffrax to use GPU computation and all possibilities offered by jax. 
Howwever, I found it to be at least 10x slower than numbalsoda in my case, and I was wondering if it was expected or I missed something. 

Here is an example code with Robertson example given in the examples. 

``` python
import numpy as np
from numba import njit
from numbalsoda import lsoda_sig, lsoda
from numba import njit, cfunc
import jax
import equinox as eqx
import jax.numpy as jnp
import diffrax
import timeit
jax.config.update("jax_enable_x64", True)

class Robertson(eqx.Module):
    k1: float
    k2: float
    k3: float

    def __call__(self, t, y, args):
        f0 = -self.k1 * y[0] + self.k3 * y[1] * y[2]
        f1 = self.k1 * y[0] - self.k2 * y[1] ** 2 - self.k3 * y[1] * y[2]
        f2 = self.k2 * y[1] ** 2
        return jnp.stack([f0, f1, f2])
    

@jax.jit
def solve_diffrax(k1, k2, k3):
    robertson = Robertson(k1, k2, k3)
    terms = diffrax.ODETerm(robertson)
    t0 = 0.0
    t1 = 100.0
    y0 = jnp.array([1.0, 0.0, 0.0])
    dt0 = 0.0002
    solver = diffrax.Kvaerno5()
    saveat = diffrax.SaveAt(ts=jnp.array([0.0, 1e-4, 1e-3, 1e-2, 1e-1, 1e0, 1e1, 1e2]))
    stepsize_controller = diffrax.PIDController(rtol=1e-8, atol=1e-8)
    sol = diffrax.diffeqsolve(
        terms,
        solver,
        t0,
        t1,
        dt0,
        y0,
        saveat=saveat,
        stepsize_controller=stepsize_controller,
    )
    return sol

sol=solve_diffrax(0.04, 3e7, 1e4)

time_taken = timeit.timeit(lambda: solve_diffrax(0.04, 3e7, 1e4), number=1000) / 1000

print("Results:")
for ti, yi in zip(sol.ts, sol.ys):
    print(f"t={ti.item()}, y={yi.tolist()}")
print(f"Took {sol.stats['num_steps']} steps in {time_taken} seconds.")



@cfunc(lsoda_sig)
def rhs(t, y, dy , k):
    dy[0] = -k[0]* y[0] + k[2] * y[1] * y[2]
    dy[1] = k[0] * y[0] - k[1] * y[1] ** 2 - k[2] * y[1] * y[2]
    dy[2] = k[1] * y[1] ** 2
    
@njit
def solve_numbalsoda(k,func):
    t_eval=np.array([0.0, 1e-4, 1e-3, 1e-2, 1e-1, 1e0, 1e1, 1e2])
    y0 = np.array([1.0, 0.0, 0.0])
    usol, success = lsoda(func, y0, t_eval, data = k,rtol=1e-8,atol=1e-8)

    return usol

func=rhs.address

k=np.array([0.04,3e7,1e4])
sol=solve_numbalsoda(k,func)

time_taken = timeit.timeit(lambda: solve_numbalsoda(k,func), number=1000) / 1000

print(sol)
print(f"Took  {time_taken} seconds.")`
```

I get 0.45ms for diffrax and 0.046ms for numbalsoda.
I will switch to diffrax anyway for the reasons explained above, but I wanted to understand why there was such a difference.

**Comment by patrick-kidger** [comment]
So you're comparing two different solvers here -- Kvaerno5 vs LSODA. That makes it hard to offer further comparison between the two libraries. For example these two methods make implicit solves in different places, and they can interpret `rtol`/`atol` in different ways. (It's not a well-known fact but the meaning of these is solver-dependent.)

FWIW when doing comparisons of this nature -- across both implementations and solver -- then a work-precision diagram is usually the best way to do this.

**Comment by jdeseze** [comment]
To add up on this question, I tried all solvers, for stiff and non-stiff problems. And I always have a factor of >10 compared to lsoda from numbalsoda - when lsoda converges, which is not always the case. 
I know it is not the same integration methods, but it is still puzzling to me why we reach such differences. Below is another simple DOE I used. Once again I'm new to diffrax so I might have missed something that slows down the computation. 

'''
import numpy as np
from numba import njit
from numbalsoda import lsoda_sig, lsoda, dop853
from numba import njit, cfunc
import jax
import equinox as eqx
import jax.numpy as jnp
import diffrax
import timeit
import matplotlib.pyplot as plt
jax.config.update("jax_enable_x64", True)

class Robertson(eqx.Module):
    k1: float
    k2: float
    k3: float

    def __call__(self, t, y, args):
        f0 = -self.k1 * y[0] + self.k3 * y[1] 
        f1 = self.k1 * y[0] - self.k2 * y[1] ** 2 - self.k3 * y[2]
        f2 = self.k2 * y[1] 
        return jnp.stack([f0, f1, f2])

@cfunc(lsoda_sig)
def rhs(t, y, dy , k):
    dy[0] = -k[0]* y[0] + k[2] * y[1] 
    dy[1] = k[0] * y[0] - k[1] * y[1] ** 2 - k[2] *  y[2]
    dy[2] = k[1] * y[1] 

func=rhs.address
k=np.array([0.04,3.,1])

# Define tolerances to test
tolerances = [1e-2, 1e-4, 1e-6, 1e-8,1e-10,1e-12]

def compute_error(sol, ref_sol):
    """Compute max absolute error between sol and ref_sol."""
    return np.linalg.norm(np.array(sol.ys) - np.array(ref_sol.ys))

# Storage for results
diffrax_errors = []
diffrax_times = []

numbalsoda_errors = []
numbalsoda_times = []

# Diffrax Solver
@jax.jit
def solve_diffrax_tol(k1, k2, k3, tol):
    robertson = Robertson(k1, k2, k3)
    terms = diffrax.ODETerm(robertson)
    t0 = 0.0
    t1 = 100.0
    y0 = jnp.array([1.0, 0.0, 0.0])
    dt0 = 0.0002
    solver = diffrax.Tsit5() 
    saveat = diffrax.SaveAt(ts=jnp.array([0.0, 1e-4, 1e-3, 1e-2, 1e-1, 1e0, 1e1, 1e2]))
    stepsize_controller = diffrax.PIDController(rtol=tol, atol=tol)
    sol = diffrax.diffeqsolve(
        terms,
        solver,
        t0,
        t1,
        dt0,
        y0,
        saveat=saveat,
        max_steps=1000000,
        stepsize_controller=stepsize_controller
        )
    return sol

# NumbaLSODA Solver
@njit
def solve_numbalsoda_tol(k, func, tol):
    t_eval = np.array([0.0, 1e-4, 1e-3, 1e-2, 1e-1, 1e0, 1e1, 1e2])
    y0 = np.array([1.0, 0.0, 0.0])
    usol, success = lsoda(
        func, y0, t_eval, data=k, rtol=tol, atol=tol
    )
    return usol


# High-precision reference solution using Diffrax
ref_tolerance = 1e-12
ref_sol = solve_diffrax_tol(0.04, 3., 1.,1e-12)

# Benchmark both solvers across tolerances
for tol in tolerances:

    # Run Diffrax and record metrics
    solve_diffrax_tol(0.04, 3., 1., tol)

    start = timeit.default_timer()
    diffrax_sol = solve_diffrax_tol(0.04, 3., 1., tol)
    diffrax_time = timeit.default_timer() - start
    diffrax_error = compute_error(diffrax_sol, ref_sol)

    diffrax_times.append(diffrax_time)
    diffrax_errors.append(diffrax_error)

    solve_numbalsoda_tol(k, func, tol)
    start = timeit.default_timer() 
    numba_sol = solve_numbalsoda_tol(k, func, tol)
    numba_time = timeit.default_timer() - start

    # Compute error (interpolate NumPy solution to match ref_sol if necessary)
    numba_error = np.max(
        np.abs(numba_sol - np.array(ref_sol.ys))
    )

    numbalsoda_times.append(numba_time)
    numbalsoda_errors.append(numba_error)

print(numbalsoda_times)
print(diffrax_times)
# Plot Work-Precision Diagram
plt.figure(figsize=(8, 6))
plt.loglog(diffrax_errors, diffrax_times, label="Diffrax Tsit5", marker="o")
plt.loglog(numbalsoda_errors, numbalsoda_times, label="NumbaLSODA", marker="s")
plt.xlabel("Error (log scale)")
plt.ylabel("Time (log scale)")
plt.title("Work-Precision Diagram")
plt.legend()
plt.grid(True, which="both", linestyle="--", linewidth=0.5)
plt.show()

'''

If it is just the solver, I'll just close the point, I just wanted to be sure I wasn't missing a big point.

**Comment by lockwo** [comment]
Make sure to use `block_until_ready()` as sometimes you can get weird timing results without it. 

Also, when I ran the original code I saw a diffrax time of 0.003 seconds and lsoda time of 1e-4 seconds (both on CPU). My remark on this is just that this sort of difference at small time scales (like 1e-4 seconds), or for small problems (few array parameters), isn't fundamentally surprising to me. I have seen my C++ implementations almost always outperform jax (not exactly the same problem, but very similar process), on small problems (I would say this is due to more careful memory management or something to that effect, but tbh idk what XLA really does with respect to memory allocations/cache optimization). Looking at their benchmarks, I see they also outperform Julia, but not when Julia is using static arrays (which is just forcing more careful/promised memory management) which are limited to very small arrays (like less than 100). If you increase the complexity of the problem (computationally more expensive vf) or want more trajectories (e.g. vmaping over the solve), then I would expect things to equalize and jax to potentially take the lead. Just my two cents, since I've been doing more C++/Jax comparison on small problems recently.

**Comment by patrick-kidger** [comment]
I've just taken a look at this and I think your benchmarking scripts are essentially correct. (Technically one should also add `jax.block_until_ready` on the output of the JAX functions, but that would only make them slower, not faster ;) )

That said... the timescales involved are very very small. I could easily believe that what is actually being measured here is actually a small amount of overhead from the Python interpreter, when entering and exiting each JIT'd function! JAX in particular does a small amount of work in Python at the start/end of each JIT'd region: checking its arguments to figure out which cached function to call, etc. I could easily believe that these have been more heavily optimized in numba. Or put another way: it isn't 10x slower, it is O(1e-4) seconds slower ;)

(Hard to say for certain without far more careful investigation -- I've not checked when LSODA switches between its stiff and nonstiff regimes on this problem, another possibility is that this is buying it a speedup, for example. I will say that we generally find Diffrax to be pretty much the same speed on real world problems as the usual options amongst Julia/Fortran/C libraries.)

**Comment by jdeseze** [comment]
Thanks for both of your answers!

My (real-world) problem has 10-15 differential equations, and 15-20 parameters, which is more complex than what I presented here but not that complex either: it is solved in ~2e-4s with numbalsoda and 4e-3s with the fastest solver and low precision with diffrax. 

To fit my parameters or perform optimization, at the end it makes a big difference in terms of time (my parameter space is quite large); however, I'm switching to diffrax for the simplicity of using vmap for my residual calculation, for the gradient calculation of my residual and to use multistart optimization. At the end the switch is valuable, but I was surprised by the difference in the solving at the beginning and wanted to be sure I wasn't missing something. 

Thanks again!

**Comment by jecampagne** [comment]
Hello, I wander if you foresee to implement the LSODA method in diffrax (I do not yet find it in the current version, have I miss something)? Thanks

**Comment by patrick-kidger** [comment]
I think we'd be open to taking a PR on it! (Or a general way of building any switching solver.)
